### ISSUE TITLE
Pointer generator

### ISSUE THREAD

**Issue by shashiongithub** [pytorch/fairseq#311]
I am trying to implement pointer-generator (See at al.) with Fairseq models. It seems to be working but then gnorm explodes. I have tried lowering the learning rate. 

Has anyone tried this before?

**Comment by huihuifan** [comment]
If you push a PR, I can take a look at the code. I implemented something
similar in fairseq.

**Comment by shashiongithub** [comment]
I have committed my changes to https://github.com/shashiongithub/fairseq .
I have not created a pull request as I was working on an older version.

**Comment by huihuifan** [comment]
Ok, I will take a look when I get a chance. Things to look out for: can the log of anything be 0 when you reconstruct the softmax with the pointer part added? Is it working correctly when there's nothing to copy? 

When you say it's working, you mean it trains for awhile and then explodes?

**Comment by shashiongithub** [comment]
Thanks Angela! I will look into your tips. 

Yes, during training it works for a while and then explodes.

**Comment by shashiongithub** [comment]
It goes something like this:

| epoch 001:     10 / 614 loss=10.519, nll_loss=10.519, ppl=1467.16, wps=1276, ups=2.4, wpb=422, bsz=17, num_updates=11, lr=0.05, gnorm=1.147, clip=100%, oom=0
| epoch 001:     20 / 614 loss=10.703, nll_loss=10.703, ppl=1666.82, wps=1469, ups=2.7, wpb=477, bsz=19, num_updates=21, lr=0.05, gnorm=0.945, clip=100%, oom=0
| epoch 001:     30 / 614 loss=10.642, nll_loss=10.642, ppl=1598.05, wps=1379, ups=2.8, wpb=450, bsz=19, num_updates=31, lr=0.05, gnorm=0.910, clip=100%, oom=0
| epoch 001:     40 / 614 loss=10.719, nll_loss=10.719, ppl=1685.17, wps=1378, ups=2.9, wpb=448, bsz=19, num_updates=41, lr=0.05, gnorm=0.860, clip=100%, oom=0
| epoch 001:     50 / 614 loss=10.702, nll_loss=10.702, ppl=1665.68, wps=1356, ups=2.9, wpb=441, bsz=18, num_updates=51, lr=0.05, gnorm=0.831, clip=100%, oom=0
| epoch 001:     60 / 614 loss=10.649, nll_loss=10.649, ppl=1605.82, wps=1345, ups=2.9, wpb=438, bsz=18, num_updates=61, lr=0.05, gnorm=0.811, clip=100%, oom=0
| epoch 001:     70 / 614 loss=10.617, nll_loss=10.617, ppl=1570.22, wps=1353, ups=2.9, wpb=442, bsz=18, num_updates=71, lr=0.05, gnorm=0.787, clip=100%, oom=0
| epoch 001:     80 / 614 loss=10.629, nll_loss=10.629, ppl=1583.45, wps=1371, ups=3.0, wpb=449, bsz=19, num_updates=81, lr=0.05, gnorm=0.766, clip=100%, oom=0
| epoch 001:     90 / 614 loss=10.582, nll_loss=10.582, ppl=1533.15, wps=1360, ups=3.0, wpb=446, bsz=18, num_updates=91, lr=0.05, gnorm=0.760, clip=100%, oom=0
| epoch 001:    100 / 614 loss=10.573, nll_loss=10.573, ppl=1523.19, wps=1370, ups=3.0, wpb=451, bsz=19, num_updates=101, lr=0.05, gnorm=0.755, clip=100%, oom=0
| epoch 001:    110 / 614 loss=10.576, nll_loss=10.576, ppl=1526.04, wps=1368, ups=3.0, wpb=449, bsz=19, num_updates=111, lr=0.05, gnorm=0.809, clip=100%, oom=0
| epoch 001:    120 / 614 loss=10.547, nll_loss=10.547, ppl=1496.53, wps=1352, ups=3.0, wpb=443, bsz=18, num_updates=121, lr=0.05, gnorm=0.827, clip=100%, oom=0
| epoch 001:    130 / 614 loss=10.520, nll_loss=10.520, ppl=1468.38, wps=1343, ups=3.0, wpb=441, bsz=18, num_updates=131, lr=0.05, gnorm=0.820, clip=100%, oom=0
| epoch 001:    140 / 614 loss=10.493, nll_loss=10.493, ppl=1440.90, wps=1326, ups=3.0, wpb=435, bsz=18, num_updates=141, lr=0.05, gnorm=0.822, clip=100%, oom=0
| epoch 001:    150 / 614 loss=10.478, nll_loss=10.478, ppl=1426.33, wps=1327, ups=3.0, wpb=436, bsz=18, num_updates=151, lr=0.05, gnorm=0.829, clip=100%, oom=0
| epoch 001:    160 / 614 loss=10.463, nll_loss=10.463, ppl=1411.57, wps=1326, ups=3.0, wpb=436, bsz=18, num_updates=161, lr=0.05, gnorm=0.823, clip=100%, oom=0
| epoch 001:    170 / 614 loss=10.449, nll_loss=10.449, ppl=1397.40, wps=1315, ups=3.0, wpb=432, bsz=18, num_updates=171, lr=0.05, gnorm=0.821, clip=100%, oom=0
| epoch 001:    180 / 614 loss=10.425, nll_loss=10.425, ppl=1374.89, wps=1316, ups=3.0, wpb=433, bsz=18, num_updates=181, lr=0.05, gnorm=0.822, clip=100%, oom=0
| epoch 001:    190 / 614 loss=10.408, nll_loss=10.408, ppl=1358.33, wps=1317, ups=3.0, wpb=434, bsz=18, num_updates=191, lr=0.05, gnorm=0.820, clip=100%, oom=0
| epoch 001:    200 / 614 loss=10.394, nll_loss=10.394, ppl=1345.99, wps=1311, ups=3.0, wpb=431, bsz=18, num_updates=201, lr=0.05, gnorm=0.817, clip=100%, oom=0
| epoch 001:    210 / 614 loss=10.370, nll_loss=10.370, ppl=1323.61, wps=1306, ups=3.0, wpb=430, bsz=17, num_updates=211, lr=0.05, gnorm=0.813, clip=100%, oom=0
| epoch 001:    220 / 614 loss=10.359, nll_loss=10.359, ppl=1312.88, wps=1303, ups=3.0, wpb=428, bsz=17, num_updates=221, lr=0.05, gnorm=0.816, clip=100%, oom=0
| epoch 001:    230 / 614 loss=10.333, nll_loss=10.333, ppl=1289.52, wps=1307, ups=3.0, wpb=429, bsz=17, num_updates=231, lr=0.05, gnorm=0.810, clip=100%, oom=0
| epoch 001:    240 / 614 loss=10.320, nll_loss=10.320, ppl=1278.44, wps=1313, ups=3.0, wpb=431, bsz=18, num_updates=241, lr=0.05, gnorm=0.807, clip=100%, oom=0
| epoch 001:    250 / 614 loss=10.300, nll_loss=10.300, ppl=1261.11, wps=1305, ups=3.0, wpb=429, bsz=18, num_updates=251, lr=0.05, gnorm=0.806, clip=100%, oom=0
| epoch 001:    260 / 614 loss=10.279, nll_loss=10.279, ppl=1242.47, wps=1302, ups=3.0, wpb=427, bsz=17, num_updates=261, lr=0.05, gnorm=0.804, clip=100%, oom=0
| epoch 001:    270 / 614 loss=10.263, nll_loss=10.263, ppl=1228.50, wps=1302, ups=3.0, wpb=428, bsz=18, num_updates=271, lr=0.05, gnorm=0.802, clip=100%, oom=0
| epoch 001:    280 / 614 loss=10.248, nll_loss=10.248, ppl=1216.36, wps=1307, ups=3.0, wpb=430, bsz=18, num_updates=281, lr=0.05, gnorm=0.799, clip=100%, oom=0
| epoch 001:    290 / 614 loss=10.232, nll_loss=10.232, ppl=1202.67, wps=1307, ups=3.0, wpb=430, bsz=18, num_updates=291, lr=0.05, gnorm=0.799, clip=100%, oom=0
| epoch 001:    300 / 614 loss=10.212, nll_loss=10.212, ppl=1185.95, wps=1310, ups=3.0, wpb=432, bsz=18, num_updates=301, lr=0.05, gnorm=0.796, clip=100%, oom=0
| epoch 001:    310 / 614 loss=10.196, nll_loss=10.196, ppl=1172.84, wps=1314, ups=3.0, wpb=433, bsz=18, num_updates=311, lr=0.05, gnorm=0.795, clip=100%, oom=0
| epoch 001:    320 / 614 loss=10.173, nll_loss=10.173, ppl=1154.53, wps=1312, ups=3.0, wpb=433, bsz=18, num_updates=321, lr=0.05, gnorm=0.793, clip=100%, oom=0
| epoch 001:    330 / 614 loss=10.161, nll_loss=10.161, ppl=1144.62, wps=1314, ups=3.0, wpb=433, bsz=18, num_updates=331, lr=0.05, gnorm=0.794, clip=100%, oom=0
| epoch 001:    340 / 614 loss=10.142, nll_loss=10.142, ppl=1130.17, wps=1318, ups=3.0, wpb=434, bsz=18, num_updates=341, lr=0.05, gnorm=0.790, clip=100%, oom=0
| epoch 001:    350 / 614 loss=10.127, nll_loss=10.127, ppl=1117.99, wps=1325, ups=3.0, wpb=436, bsz=18, num_updates=351, lr=0.05, gnorm=0.789, clip=100%, oom=0
| epoch 001:    360 / 614 loss=10.107, nll_loss=10.107, ppl=1102.77, wps=1332, ups=3.0, wpb=438, bsz=18, num_updates=361, lr=0.05, gnorm=0.788, clip=100%, oom=0
| epoch 001:    370 / 614 loss=10.093, nll_loss=10.093, ppl=1092.29, wps=1334, ups=3.0, wpb=439, bsz=18, num_updates=371, lr=0.05, gnorm=0.788, clip=100%, oom=0
| epoch 001:    380 / 614 loss=10.071, nll_loss=10.071, ppl=1075.98, wps=1337, ups=3.0, wpb=441, bsz=18, num_updates=381, lr=0.05, gnorm=0.788, clip=100%, oom=0
| epoch 001:    390 / 614 loss=10.053, nll_loss=10.053, ppl=1062.59, wps=1342, ups=3.0, wpb=443, bsz=18, num_updates=391, lr=0.05, gnorm=0.787, clip=100%, oom=0
| epoch 001:    400 / 614 loss=10.034, nll_loss=10.034, ppl=1048.57, wps=1341, ups=3.0, wpb=442, bsz=18, num_updates=401, lr=0.05, gnorm=0.788, clip=100%, oom=0
| epoch 001:    410 / 614 loss=10.027, nll_loss=10.027, ppl=1043.22, wps=1340, ups=3.0, wpb=442, bsz=18, num_updates=411, lr=0.05, gnorm=0.794, clip=100%, oom=0
| epoch 001:    420 / 614 loss=10.012, nll_loss=10.012, ppl=1032.65, wps=1340, ups=3.0, wpb=441, bsz=18, num_updates=421, lr=0.05, gnorm=0.795, clip=100%, oom=0
| epoch 001:    430 / 614 loss=9.997, nll_loss=9.997, ppl=1021.81, wps=1340, ups=3.0, wpb=442, bsz=18, num_updates=431, lr=0.05, gnorm=11.481, clip=100%, oom=0
| epoch 001:    440 / 614 loss=9.979, nll_loss=9.979, ppl=1008.86, wps=1345, ups=3.0, wpb=444, bsz=18, num_updates=441, lr=0.05, gnorm=13.519, clip=100%, oom=0
| epoch 001:    450 / 614 loss=9.962, nll_loss=9.962, ppl=997.40, wps=1344, ups=3.0, wpb=444, bsz=18, num_updates=451, lr=0.05, gnorm=22.777, clip=100%, oom=0
| epoch 001:    460 / 614 loss=9.950, nll_loss=9.950, ppl=989.20, wps=1340, ups=3.0, wpb=442, bsz=18, num_updates=461, lr=0.05, gnorm=59.928, clip=100%, oom=0
| epoch 001:    470 / 614 loss=9.943, nll_loss=9.943, ppl=984.62, wps=1341, ups=3.0, wpb=443, bsz=18, num_updates=471, lr=0.05, gnorm=107.522, clip=100%, oom=0
| epoch 001:    480 / 614 loss=9.949, nll_loss=9.949, ppl=988.45, wps=1342, ups=3.0, wpb=443, bsz=18, num_updates=481, lr=0.05, gnorm=291.718, clip=100%, oom=0
| epoch 001:    490 / 614 loss=9.974, nll_loss=9.974, ppl=1005.83, wps=1342, ups=3.0, wpb=443, bsz=18, num_updates=491, lr=0.05, gnorm=876.141, clip=100%, oom=0
| epoch 001:    500 / 614 loss=9.990, nll_loss=9.990, ppl=1016.65, wps=1341, ups=3.0, wpb=443, bsz=18, num_updates=501, lr=0.05, gnorm=1035.875, clip=100%, oom=0
| epoch 001:    510 / 614 loss=10.012, nll_loss=10.012, ppl=1032.58, wps=1339, ups=3.0, wpb=443, bsz=18, num_updates=511, lr=0.05, gnorm=1957.003, clip=100%, oom=0
| epoch 001:    520 / 614 loss=10.043, nll_loss=10.043, ppl=1055.29, wps=1344, ups=3.0, wpb=444, bsz=18, num_updates=521, lr=0.05, gnorm=3575.989, clip=100%, oom=0
